internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -31,34 +32,91 @@ type contextProviderK8sSecrets struct {
 	logger *logger.Logger
 	config *Config
 
-	clientMx sync.Mutex
-	client   k8sclient.Interface
+	clientMx    sync.Mutex
+	initialized sync.WaitGroup
+
+	k8sCacheProvider cacheProvider
+
+	// readers holds a set of cached K8S reader per namespace.
+	// We restrict each cache to watch only 1 namespace. Watching all the namespaces would consume a lot of memory
+	// when we probably want to watch a very limited set of Secrets.
+	readers map[string]client.Reader
+
+	// ctx is the context used to start the informers. Informers will be stopped once the context is done.
+	ctx context.Context
+}
+
+type cacheProvider interface {
+	new(cfg *Config, namespace string) (cache.Cache, error)
+}
+
+type defaultCacheProvider struct{}
+
+func (dcp defaultCacheProvider) new(cfg *Config, namespace string) (cache.Cache, error) {
+	restConfig, err := kubernetes.BuildConfig(cfg.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	if cfg.KubeClientOptions.QPS > 0 {
+		restConfig.QPS = cfg.KubeClientOptions.QPS
+	}
+	if cfg.KubeClientOptions.Burst > 0 {
+		restConfig.Burst = cfg.KubeClientOptions.Burst
+	}
+	return cache.New(restConfig, cache.Options{
+		DefaultNamespaces: map[string]cache.Config{namespace: {}},
+	})
+}
+
+// getReader returns a cached client associated to a given Kubernetes namespace.
+func (p *contextProviderK8sSecrets) getReader(namespace string) (client.Reader, error) {
+	p.initialized.Wait()
+	p.clientMx.Lock()
+	defer p.clientMx.Unlock()
+	reader, exist := p.readers[namespace]
+	if !exist {
+		p.logger.Infof(
+			"Create new cached reader for namespace %s",
+			namespace,
+		)
+		newReader, err := p.k8sCacheProvider.new(p.config, namespace)
+		if err != nil {
+			return nil, err
+		}
+		p.readers[namespace] = newReader
+		reader = newReader
+		go func() {
+			if err := newReader.Start(p.ctx); err != nil {
+				p.logger.Errorf("Could not start K8s client reader: %v", err)
+			}
+		}()
+		// Wait for the cache to be initialized.
+		newReader.WaitForCacheSync(p.ctx)
+	}
+	return reader, nil
 }
 
 // ContextProviderBuilder builds the context provider.
-func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed bool) (corecomp.ContextProvider, error) {
+func ContextProviderBuilder(logger *logger.Logger, c *config.Config, _ bool) (corecomp.ContextProvider, error) {
```

Comment:
I assume that this change has the Elastic Agent now watching for changes in secrets, it would be great to see this also now trigger a re-render of the policy when a secret changes. Seems that the code is already listening to the API for those events, why not take it to the next level and improve it some more?

Reply:
> it would be great to see this also now trigger a re-render of the policy when a secret changes.

I’m afraid I don’t know what "trigger a re-render of the policy" means.

Reply:
See my comment above for more context. It means that when a secret value is updated that you would call to the `comm Communicator` that a value has changed. That will signal the Elastic Agent composable manager to re-render the component model because something changed.

Think of it like this:

1. Elastic Agent is running with a policy and inside of that policy it references a password from a kubernetes secret.
2. The kubernetes secret is updated to change the password.
3. Elastic Agent gets the event that the password has changed.
4. It updates its cache for the secret.
5. It alerts the composable manager that a value has changed.
6. The Elastic Agent re-renders the policy with that new password substituted.
7. The Elastic Agent sends the new configuration to the running component.
8. The running component starts using the new password.

Elastic Agent already works that way today, just not for kubernetes secrets (everything else does). The kubernetes secrets provider just needs to "5." in that list and Elastic Agent would do the rest for you.

Reply:
> I assume that this change has the Elastic Agent now watching for changes in secrets,

The benefit of using the controller runtime client is to get an abstraction over the watch mechanism, and therefore to get a eventually up-to-date cache without managing watches. Given your explanations, I feel less and less confident in continuing down this path.